libbpf-tools/futexctn.bpf.c

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "futexctn.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES	10240
+
+#define FUTEX_WAIT             0
+#define FUTEX_PRIVATE_FLAG     128
+#define FUTEX_CLOCK_REALTIME   256
+#define FUTEX_CMD_MASK         ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)
+
+const volatile bool targ_per_stack = false;
+const volatile bool targ_ms = false;
+const volatile pid_t targ_pid = 0;
+const volatile pid_t targ_tid = 0;
+
+struct val_t {
+	u64 ts;
+	u64 uaddr;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, struct val_t);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct hist_key);
+	__type(value, struct hist);
+} hists SEC(".maps");
+
+static struct hist initial_hist;
+
+SEC("tracepoint/syscalls/sys_enter_futex")
+int futex_enter(struct trace_event_raw_sys_enter *ctx)
+{
+	struct val_t v = {};
+	u64 pid_tgid;
+	u32 tid;
+
+	if (((int)ctx->args[1] & FUTEX_CMD_MASK) != FUTEX_WAIT)
+		return 0;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	tid = (__u32)pid_tgid;
+	if (targ_pid && targ_pid != pid_tgid >> 32)
+		return 0;
+	if (targ_tid && targ_tid != tid)
+		return 0;
+	v.uaddr = ctx->args[0];
+	v.ts = bpf_ktime_get_ns();
+	bpf_map_update_elem(&start, &pid_tgid, &v, 0);
+	return 0;
+}
+
+SEC("tracepoint/syscalls/sys_exit_futex")
+int futex_exit(struct trace_event_raw_sys_enter *ctx)
```

Comment:
The argument type does not sounds right. Should it be `trace_event_raw_sys_exit`?

```
struct trace_event_raw_sys_exit {
        struct trace_entry ent;
        long id;
        long ret;       
        char __data[0]; 
};
```

Also, since we have 'ret' value here, should we check whether the syscall is successfully or not, for either WAIT or WAKE? We should only care successful case, right?

Reply:
> The argument type does not sounds right. Should it be trace_event_raw_sys_exit?

Yes, it should be `trace_event_raw_sys_exit`.

> Also, since we have 'ret' value here, should we check whether the syscall is successfully or not, for either WAIT or WAKE? We should only care successful case, right?

Agree, successful case is more make sence.

@yonghong-song Both are fixed, thanks!